src/code/sched.c

```diff
@@ -165,90 +227,149 @@ OSScTask* func_800C89D4(SchedContext* sc, OSScTask* task) {
         sc->audioListHead = sc->audioListHead->next;
         if (sc->audioListHead == NULL) {
             sc->audioListTail = NULL;
         }
     } else if (gfxTask != NULL) {
-        if (gfxTask->state & OS_SC_YIELDED || !(gfxTask->flags & OS_SC_NEEDS_RDP)) {
-            if (ret & OS_SC_SP) {
-                *sp = gfxTask;
-                ret &= ~OS_SC_SP;
+        // GFX Task
+
+        if ((gfxTask->state & OS_SC_YIELDED) || !(gfxTask->flags & OS_SC_NEEDS_RDP)) {
+            // If this is a yielded GFX task, or the RDP is not needed for this GFX task
+
+            if (state & OS_SC_SP) {
+                // If the RSP is available, return next graphics task
+
+                *spTaskOut = gfxTask;
+                // RSP required
+                nextState &= ~OS_SC_SP;
+                // Advance task queue
                 sc->gfxListHead = sc->gfxListHead->next;
                 if (sc->gfxListHead == NULL) {
                     sc->gfxListTail = NULL;
                 }
             }
-        } else if (ret == (OS_SC_SP | OS_SC_DP)) {
-            if (gfxTask->framebuffer == NULL || func_800C89D4(sc, gfxTask) != NULL) {
-                *sp = *dp = gfxTask;
-                ret &= ~(OS_SC_SP | OS_SC_DP);
+        } else if (state == (OS_SC_SP | OS_SC_DP)) {
+            // Both the RSP and RDP are available, check requested framebuffer
+            if (gfxTask->framebuffer == NULL || Sched_GfxTaskFramebufferValid(sc, gfxTask) != NULL) {
+                // Return next graphics task
+                *spTaskOut = *dpTaskOut = gfxTask;
+                // RSP and RDP both required
+                nextState &= ~(OS_SC_SP | OS_SC_DP);
+                // Advance task queue
                 sc->gfxListHead = sc->gfxListHead->next;
                 if (sc->gfxListHead == NULL) {
                     sc->gfxListTail = NULL;
                 }
             }
         }
     }
-    return ret;
+    return nextState;
 }
 
-void func_800C8BC4(SchedContext* sc, OSScTask* task) {
+/**
+ * Sets the next framebuffer to the framebuffer associated to `task`.
+ * If there is no current buffer or it is time to swap, this buffer will be swapped to
+ * immediately, otherwise it will be swapped to later in Sched_HandleRetrace.
+ *
+ * @see Sched_HandleRetrace
+ */
+void Sched_SetNextFramebufferFromTask(Scheduler* sc, OSScTask* task) {
     if (sc->pendingSwapBuf1 == NULL) {
         sc->pendingSwapBuf1 = task->framebuffer;
 
         LogUtils_CheckValidPointer("sc->pending_swapbuffer1", sc->pendingSwapBuf1, "../sched.c", 618);
 
-        if ((sc->curBuf == NULL) || (sc->curBuf->updateRate2 < 1)) {
-            func_800C84E4(sc, task->framebuffer);
+        if (sc->curBuf == NULL || sc->curBuf->updateTimer <= 0) {
+            Sched_SwapFrameBuffer(sc, task->framebuffer);
         }
     }
 }
 
-u32 Sched_IsComplete(SchedContext* sc, OSScTask* task) {
+/**
+ * Checks if the task is done, i.e. it is no longer running on either the RSP and RDP.
```

Comment:
```suggestion
 * Checks if the task is done, i.e. it is no longer running on either the RSP or RDP.
```
not sure if it should be "either ... or" or "both ... and"